app/services/gps_report.rb

```diff
@@ -0,0 +1,131 @@
+
+    {
+      failures: failures,
+      move_count: moves.count,
+    }
+  end
+
+private
+
+  def failures
+    return @failures unless @failures.nil?
+
+    @failures = moves.filter_map { |move|
+      { move: move, reason: 'no_journeys' } if move.journeys.empty?
+    }.concat(journey_failures)
+  end
+
+  def journey_failures
+    failures = journeys.filter_map do |journey|
+      gps_data = gps_data_map[journey.id]
+      next { move: journey.move, reason: 'no_gps_data' } if gps_data.blank?
+
+      intervals = gps_data.each_with_index.filter_map { |t, i| i < gps_data.count - 1 ? gps_data[i + 1] - t : nil }
+      { move: journey.move, reason: 'gps_data_gap' } if intervals.max > 60.seconds
+    end
+    failures.uniq { |f| f[:move] }
+  end
+
+  def gps_data_map
+    return @gps_data_map unless @gps_data_map.nil?
+
+    repair_table
+
+    Rails.logger.info "Fetching #{@supplier.key} GPS data points."
+
+    time_since = TimeSince.new
+    res = athena_query("select journey_id, tracking_timestamp from #{@supplier.key} where journey_id in ('#{journeys.map(&:id).join("', '")}') order by journey_id, tracking_timestamp;")
+
+    @gps_data_map = {}
+    res.result_set.rows[1..].each do |data_row|
+      journey_id, timestamp = *data_row.data.map(&:var_char_value)
+      @gps_data_map[journey_id] = [] if @gps_data_map[journey_id].nil?
+
+      @gps_data_map[journey_id] << Time.zone.parse(timestamp)
+    end
+
+    Rails.logger.info "Fetched #{pluralize(res.result_set.rows[1..].count, "#{@supplier.key} GPS data point")} in #{pluralize(time_since.get, 'second')}."
+
+    @gps_data_map
+  end
+
+  def journeys
+    @journeys ||= moves.flat_map(&:journeys)
+  end
+
+  def moves
+    return @moves unless @moves.nil?
+
+    @moves = []
+    Move.where(status: 'completed', date: @from_date..@to_date, supplier: @supplier).order(:id).find_each do |move|
+      @moves << move
+    end
+
+    @moves
+  end
+
+  def repair_table
+    Rails.logger.info "Repairing table `#{@supplier.key}`."
+
+    time_since = TimeSince.new
+    athena_query("MSCK REPAIR TABLE #{@supplier.key};")
+
+    Rails.logger.info "#{@supplier.key} table repair took #{pluralize(time_since.get, 'second')}."
+  end
+
+  def athena_query(query)
+    query_id = athena_client.start_query_execution({
+      work_group: ENV['ATHENA_WORK_GROUP'],
+      query_execution_context: { database: ENV['ATHENA_DATABASE'] },
+      query_string: query,
+    }).query_execution_id
+
+    loop do
+      sleep 1.second
+
+      begin
+        return athena_client.get_query_results(query_execution_id: query_id)
+      rescue Aws::Athena::Errors::InvalidRequestException => e
+        unless e.message.starts_with?('Query has not yet finished')
+          Rails.logger.error athena_client.get_query_execution(query_execution_id: query_id)
+
+          raise e
+        end
+      end
+    end
```

Comment:
I think we can use [`get_query_executation`](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Athena/Client.html#get_query_execution-instance_method) to check the state of the query, which might be better than this loop (it protects us if the error message changes).

I'm imagining something like this:

```suggestion
    while %w[QUEUED RUNNING].include?(athena_client.get_query_execution(query_execution_id: query_id).status.state)
      sleep 1.second
    end
      
    return athena_client.get_query_results(query_execution_id: query_id)
```